man/man8/monitorix.8
.\" Monitorix manpage.
.\" Copyright (C) 2005-2024 by Jordi Sanfeliu <jordi@fibranet.cat>
.\"
.\" This is the man page for the monitorix collector daemon.
.\"
.TH monitorix 8 "Nov 2024" 3.16.0 "Monitorix collector daemon"
.SH NAME
monitorix - a lightweight system monitoring tool
.SH SYNOPSIS
\fBmonitorix\fR \fB-c\fR configfile [\fB-p\fR pidfile] [\fB-d\fR none | graph[,graph] | all] [\fB-v\fR] [\fB-n\fR] [\fB-u\fR] [\fB-s\fR splitpolicy] [\fB-e\fR report=<timeframe>,graphs=<graph>[+graph][+graph...],to=<email>]
.SH DESCRIPTION
This Perl daemon starts the main Monitorix process, which gathers statistics about the system it is running on and stores this information in a set of RRD files.
.SH OPTIONS
.TP
\fB\-c\fR \fIconfigfile\fR
The default location of the main configuration file varies depending on the operating system:
.P
.RS
Linux:        \fI/etc/monitorix/monitorix.conf\fP
.br
FreeBSD:      \fI/usr/local/etc/monitorix.conf\fP
.P
Extra configuration files can be placed in \fI/etc/monitorix/conf.d\fP, they will be loaded right after the main configuration file (overriding previous options).
.RE
.TP
\fB\-p\fR \fIpidfile\fR
Stores the daemon's process ID into the specified file.
.TP
\fB\-d\fR none | graph[,graph] | all
Logs more information about what \fBmonitorix\fP is doing internally. The keys reflect if the data collected must be displayed on each case. If \fBall\fP is defined, the data collected of all enabled graphs will be shown. If \fBnone\fP is defined no data collected will be shown. Finally it is also possible to define a comma-separated list of graph names from which show their data collected.
.br
For a reference of all graph names check the option \fBgraph_name\fP in the \fImonitorix.conf\fP configuration file.
.TP
\fB\-v\fR
Displays version information.
.TP
\fB\-n\fR
This argument will prevent Monitorix from daemonizing, and hence, forcing it to run in foreground. This is specially useful for debugging purposes.
.TP
\fB\-u\fR
This option will force Monitorix to run under the regular user who started it. It's necessary to make sure that this user will have write permissions to the directory and files pointed by the options \fBbase_dir\fP, \fBbase_lib\fP and \fBlog_file\fP (either for \fImonitorix\fP and for the HTTP built-in server). Also, you must know that some graphs might not work because only the 'root' user is capable to get such statistics. Check the log files after starting Monitorix in this mode.
.TP
\fB\-s\fR \fIsplitpolicy\fR
This option decides which part of a line in the config file will be the key and which one will be the value. The split policy accepts the values \fIguess\fP (which is the default), \fIwhitespace\fP (which causes the Monitorix to split by whitespace) and \fIequalsign\fP (which causes it to split strictly by equal sign).
.TP
\fB\-e\fR report=\fItimeframe\fR,graphs=\fIgraph\fR[\fI+graph\fR][\fI+graph\fR...],to=\fIemail\fR
This option permits to send email reports unscheduled (i.e: at any time). You don't need to stop your current Monitorix instance, just execute a new one with the new parameter -e and once the job is done it will terminate itself, without affecting your current Monitorix instance. The following is an example of how to send an email report that includes the last day of \fIsystem\fR, \fIkernel\fR and \fIhptemp\fR graphs to the \fIuser@example.com\fR email address:
.P
.RS
# monitorix -c </etc/monitorix/conf.d/local.conf> -e report=daily,graphs=system+kern+hptemp,to=user@example.com
.P
This option requires to have previously enabled the \fIemailreports\fR module.
.RE
.SH SIGNALS
On receipt of a SIGHUP, \fBmonitorix\fP will close and reopen its log file (provided that it has a filename defined). This is useful in scripts which rotate and age log files.
.P
Note that the configuration file is not re-read.
.SH FILES
The following set of the files are created on every startup:
.TP 8
\fB<base_dir>/cgi/monitorix.conf.path\fP
Stores the path of \fIconfigfile\fP and other information.
.TP 8
\fB<base_dir>/cgi/monitorix.hplog\fP
Stores the output of \fIhplog\fP command.
.TP 8
\fB<base_dir>/index.html\fP
HTML main page.
.SH AUTHOR
Monitorix is written by Jordi Sanfeliu <jordi@fibranet.cat>
.SH COPYRIGHT
Copyright \(co 2005-2024 Jordi Sanfeliu
.br
Licensed under the GNU General Public License version 2 (GPLv2).
.SH "SEE ALSO"
.BR monitorix.conf (5)


Changes
3.16.0 - 27-Nov-2024
====================
- Added support for MongoDB version 6 and changed the way how 'mongodb.pm' is
  configured. [#451]
- Added the new option 'log_successful_requests' in the HTTP built-in server, to
  be able to not logging successful requests (enabled by default). [#454]
- Changed the default value of 'global_zoom' so now all graphs are 50% bigger.
- Adjusted the .spec file to be able to generate an RPM file to be installed on
  systemd or SysV init systems. [#462]
- Removed the following obsolete options:'secure_log', 'secure_log_date_format',
  'imap_log', 'imap_log_date_format', 'hylafax_log', 'cups_log' and 'cg_logdir'.
- Fixed the size of graphs in Multihost viewer when using a bigger 'global_zoom'
  value.
- Fixed the colors based on the meaning of input (upload) and output (download)
  from the server point of view in 'ftp.pm'.
- Fixed the error message 'Use of uninitialized value $gen_h[5] in join or
  string at /usr/lib/monitorix/mail.pm line 668'.
- Fixed the regexp that retrieves the reason for last transfer to battery string
  in 'nut.pm'. [#466]
- Fixed the update() function to support newer versions of Tinyproxy.
- Fixed to LINE1 the Involuntary Context Switches value in 'process.pm'.


3.15.0 - 07-Dec-2022
====================
- Added a complete graph to monitor the power consumption of Intel-RAPL capable
  devices ('intelrapl.pm'). [#390]
- Added a complete graph to monitor AMD CPU power consumption ('amdenergy.pm').
  [#389]
- Complete rewrite of the 'serv.pm' module making it fully customizable. [#115]
- Added the 'additional_graph_name' configuration option. [#388]
- Added support to be able to rename ipmi sensors that include whitespaces in
  their name. [#393]
- Added the option 'use_nan_for_missing_data' to 'du.pm' and 'fs.pm'. [#399]
- Added some new options to handle zero values as NaN in 'nut.pm'. [#401]
- Added a new option 'enable_rrd_lock' (disabled by default) to synchronise the
  rrd file access. [#370]
- Added the option 'use_nan_for_missing_data' to 'lmsens.pm'. [#406]
- Added support for the 'mV' unit to voltage keys in 'lmsens.pm'. [#315]
- Added some new options to (optionally) change scaling in 'nut.pm'. [#409]
- Added support to enable relative URLs in graphs. [#410]
- Added more accuracy when measuring the time elapsed between updates, instead
  of assuming always 60 seconds (in 'fs.pm', 'libvirt.pm', 'net.pm', 'port.pm'
  and 'process.pm'). [#407]
- Added some graph enhancements in 'nut.pm'. [#409]
- Added the option 'gap_on_all_nan' in 'du.pm' and 'fs.pm'. [#417]
- Added the new option 'has_removable_devices' (disabled by default) in 'fs.pm'
  to recalculate the device name of each filesystem on every update. [#418]
- Added support for AMaViS Module Mail::SPF. [#419]
- Added option to disable rrd locking during cgi calls. [#428]
- Added CSS theming support in 'emailreports.pm'.
- Added the -e parameter in 'monitorix' to be able to send emailreports at any
  time. [#322][#429]
- Added the option 'respect_power_state' in 'amdgpu.pm' to avoid waking up GPUs
  in D3 power state. [#433]
- Included more NVIDIA variants of 'N/A' in 'nvidiagpu.pm'. [#415]
- Changed to '/usr/bin/vcgencmd' the default path in 'monitorix.conf'. [#420]
- Updated 'apcupsd.pm' to have the same options and fixes as 'nut.pm'. [#440]
- Improved the readability of the Context switches graph and value alignments
  in 'process.pm'. [#442][#443]
- Removed a call to print() inside of push() that displayed the return value
  (1) when monitoring multiple BIND servers.
- Fixed an error in 'pgsql.pm' if a non-privileged user is used to read the
  statistics. [#384]
- Fixed the amdgpu section in 'monitorix.conf'. [#385]
- Fixed a change in the quotation marks introduced in 3.14 that prevented from
  sending reports. [#391]
       (thanks to Michael Bronzini, mb20 AT bronzeware.se for pointing this out)
- Fixed wrong index for image directory in 'amdgpu.pm', 'nvidiagpu.pm' and
  'nvme.pm'. [#387]
- Fixed the extra padding in 'ipmi.pm' and 'du.pm'. [#392]
- Fixed the CSS theming support in the bottom of some graphs. [#405]
- Fixed graphs layout when using more than one server in 'chrony.pm'. [#408]
- Fixed the legend alignment in 'nvidiagpu.pm' and 'amdgpu.pm'. [#411]
- Fixed Makefile to do a backup of the current 'monitorix.conf'. [#414]
- Fixed to match the values in the zoom graph of cores in 'lmsens.pm'.
- Fixed warnings for undefined mimetype on some HTTP requests. [#427]
- Small fixes and cosmetic changes.


3.14.0 - 18-Jan-2022
====================
- Added a complete graph to support NVMe device statistics ('nvme.pm'). [#215]
- Added a complete graph to support AMD graphic cards statistics ('amdgpu.pm').
  [#367]
- Added a complete graph to support NVIDIA graphic cards with more extended
  statistics ('nvidiagpu.pm'). [#333]
- Added support in 'redis.pm' to connect to a socket file. [#316]
- Added support to get temperature values for gpu[n] keys using the output of
  lm_sensors in 'lmsens.pm'. [#320]
- Added support in 'process.pm' for systems with different PID max value defined
  in '/proc/sys/kernel/pid_max'.
- Added new graph of process' uptime in 'process.pm'. It comes with a new option
  called 'time_unit' to set the scale of the graph. [#311]
- Added the option 'time_unit' in 'system.pm' to set the scale of the uptime
  graph. [#331]
- Added the new command line option '-s' to decide which part of a line in the
  config file will be the key and which one will be the value. The split policy
  accepts the values 'guess' (as the default), 'whitespace' and 'equalsign'.
	(suggested by Shashi Mall, shashi.mall AT wizix.com)
- Added support for IPv6 addresses in 'traffacct.pm'.
	(thanks to Adam Starr, astar AT fhtc.edu)
- Added the option 'cmd' in 'lmsens.pm', which defaults to 'sensors', to be able
  to help to add sensors values not covered by lm_sensors. [#325]
- Added the ability to prefix fan values to be able to select between rpm and

lib/Monitorix.pm
#
# Monitorix - A lightweight system monitoring tool.
#
# Copyright (C) 2005-2020 by Jordi Sanfeliu <jordi@fibranet.cat>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#

package Monitorix;

use strict;
use warnings;
use Exporter 'import';
use POSIX qw(setuid setgid setsid getgid getuid ceil);
use Fcntl qw(:flock);
use Socket;
our @EXPORT = qw(logger trim min max celsius_to img_element picz_a_element picz_js_a_element uptime2str setup_riglim httpd_setup get_nvidia_data get_ati_data flush_accounting_rules lockfile_handler global_flock);

sub logger {
	my ($msg) = @_;

	$msg = localtime() . " - " . $msg;
	print("$msg\n");
}

sub trim {
	my $str = shift;

	if($str) {
		$str =~ s/^\s+//;
		$str =~ s/\s+$//;
		return $str;
	}
}

sub min {
	my ($min, @args) = @_;
	foreach(@args) {
		$min = $_ if $_ < $min;
	}
	return $min;
}

sub max {
	my ($max, @args) = @_;
	foreach(@args) {
		$max = $_ if $_ > $max;
	}
	return $max;
}

sub celsius_to {
	my ($config, $celsius) = @_;

	$celsius = $celsius || 0;
	if(lc($config->{temperature_scale}) eq "f") {
		return ($celsius * (9 / 5)) + 32;
	}
	return $celsius;
}

sub img_element {
	my %params = @_;
	return sprintf '<img src="%s/%s%s" border="0">',
					 $params{config}->{url},
					 $params{config}->{imgs_dir},
					 $params{IMG};
}

sub picz_a_element {
	my %params = @_;
	return sprintf '<a href="%s/%s%s">%s</a>',
					 $params{config}->{url},
					 $params{config}->{imgs_dir},
					 $params{IMGz},
					 img_element(%params);

}

sub picz_js_a_element {
	my %params = @_;

	my $zoom = (uc($params{config}->{image_format}) eq "SVG") ? (4 / 3) : 1;

	my $js = sprintf "javascript:void(window.open('%s/%s%s','','width=%d,height=%d,scrollbars=0,resizable=0'))",
					   $params{config}->{url},
					   $params{config}->{imgs_dir},
					   $params{IMGz},
					   ceil($params{width} * $zoom),
					   ceil($params{height} * $zoom + 0.5);

	return sprintf '<a href="%s">%s</a>',
					 $js,
					 img_element(%params);
}

sub uptime2str {
	my $uptime = shift || 0;
	my $str;

	my $d = int($uptime / (60 * 60 * 24));
	my $h = int($uptime / (60 * 60)) % 24;
	my $m = int($uptime / 60) % 60;

	my $d_string = $d ? sprintf("%d days,", $d) : "";
	my $h_string = $h ? sprintf("%d", $h) : "";
	my $m_string = $h ? sprintf("%sh %dm", $h, $m) : sprintf("%d min", $m);

	return "$d_string $m_string";
}

sub setup_riglim {
	my $myself = (caller(0))[3];
	my ($rigid, $limit) = @_;
	my @riglim;

	my ($upper, $lower) = split(':', trim($limit) || "0:0");
	if(trim($rigid || 0) eq 0) {
		push(@riglim, "--lower-limit=$lower") if defined($lower);
	} else {
		push(@riglim, "--upper-limit=" . ($upper || 0));
		push(@riglim, "--lower-limit=" . ($lower || 0));
		push(@riglim, "--rigid") if trim($rigid || 0) eq 2;
	}
	return \@riglim;
}

sub httpd_setup {
	my $myself = (caller(0))[3];
	my ($config, $reguser) = @_;
	my $pid;
	my ($uid, $gid);

	my $host = $config->{httpd_builtin}->{host};
	my $port = $config->{httpd_builtin}->{port};

	if($reguser) {
		(undef, undef, $uid, $gid ) = getpwuid($<);
	} else {
		(undef, undef, $uid) = getpwnam($config->{httpd_builtin}->{user});
		(undef, undef, $gid) = getgrnam($config->{httpd_builtin}->{group});
	}

	if(!defined($uid)) {
		logger("$myself: ERROR: invalid user defined.");
		return;
	}
	if(!defined($gid)) {
		logger("$myself: ERROR: invalid group defined.");
		return;
	}
	if(!defined($port)) {
		logger("$myself: ERROR: invalid port defined.");
		return;
	}

	if($pid = fork()) {
		$config->{httpd_pid} = $pid;
		return;	# parent returns
	}

	# create the HTTPd logfile
	if($config->{log_file}) {
		open(OUT, ">> " . $config->{httpd_builtin}->{log_file});
		close(OUT);
		chown($uid, $gid, $config->{httpd_builtin}->{log_file});
		chmod(0600, $config->{httpd_builtin}->{log_file});
	}

	setgid($gid);
	if(getgid() != $gid) {
		logger("WARNING: $myself: unable to setgid($gid).");
		exit(1);
	}
	setuid($uid);
	if(getuid() != $uid) {
		logger("WARNING: $myself: unable to setuid($uid).");
		exit(1);
	}
	setsid();
	$SIG{$_} = 'DEFAULT' for keys %SIG;		# reset all sighandlers
	$0 = "monitorix-httpd listening on $port";	# change process' name
	chdir($config->{base_dir});

	# check if 'htpasswd' file does exists and it's accessible
	if(lc($config->{httpd_builtin}->{auth}->{enabled}) eq "y") {
		if(! -r ($config->{httpd_builtin}->{auth}->{htpasswd} || "")) {
			logger("$myself: '$config->{httpd_builtin}->{auth}->{htpasswd}' $!");
		}
	} else {
		if(!grep {$_ eq $config->{httpd_builtin}->{host}} ("localhost", "127.0.0.1")) {
			logger("WARNING: the HTTP built-in server has authentication disabled.");
		}
	}

	my $server = HTTPServer->new();
	$server->host($host);
	$server->port($port);
	$server->run();
	exit(0);
}

sub get_nvidia_data {
	my $myself = (caller(0))[3];
	my ($gpu) = @_;
	my $total = 0;
	my $used = 0;
	my $mem = 0;
	my $cpu = 0;
	my $temp = 0;
	my $check_mem = 0;
	my $check_cpu = 0;
	my $check_temp = 0;
	my $l;

	my @data = ();
	if(open(IN, "nvidia-smi -q -i $gpu -d MEMORY,UTILIZATION,TEMPERATURE |")) {
		@data = <IN>;
		close(IN);
	} else {
		logger("$myself: ERROR: 'nvidia-smi' command is not installed.");
	}
	for($l = 0; $l < scalar(@data); $l++) {
		if($data[$l] =~ /Memory Usage/) {
			if($data[$l] !~ /BAR1 Memory Usage/) {
				$check_mem = 1;
				next;
			}
		}
		if($check_mem) {
			if($data[$l] =~ /Total/) {
				my (undef, $tmp) = split(':', $data[$l]);
				if($tmp eq "\n") {
					$l++;
					$tmp = $data[$l];
				}
				my ($value, undef) = split(' ', $tmp);
				$value =~ s/[-]/./;
				$value =~ s/[^0-9.]//g;
				if(int($value) > 0) {
					$total = int($value);
				}
			}
			if($data[$l] =~ /Used/) {
				my (undef, $tmp) = split(':', $data[$l]);
				if($tmp eq "\n") {
					$l++;
					$tmp = $data[$l];
				}
				my ($value, undef) = split(' ', $tmp);
				$value =~ s/[-]/./;
				$value =~ s/[^0-9.]//g;
				if(int($value) > 0) {
					$used = int($value);
				}
				$check_mem = 0;
			}
		}

		if($data[$l] =~ /Utilization/) {
			$check_cpu = 1;
			next;
		}
		if($check_cpu) {
			if($data[$l] =~ /Gpu/) {
				my (undef, $tmp) = split(':', $data[$l]);
				if($tmp eq "\n") {
					$l++;
					$tmp = $data[$l];
				}
				my ($value, undef) = split(' ', $tmp);
				$value =~ s/[-]/./;
				$value =~ s/[^0-9.]//g;
				$value ||= 0;	# zero if not numeric
				if(int($value) > 0) {
					$cpu = int($value);
				}
			}
			# not used
			if($data[$l] =~ /Memory/) {
				my (undef, $tmp) = split(':', $data[$l]);
				if($tmp eq "\n") {
					$l++;
					$tmp = $data[$l];
				}
				my ($value, undef) = split(' ', $tmp);
				$value =~ s/[-]/./;
				$value =~ s/[^0-9.]//g;
				$value ||= 0;	# zero if not numeric
				if(int($value) > 0) {
					$mem = int($value);
				}
			}
			$check_cpu = 0;
		}

		if($data[$l] =~ /Temperature/) {
			$check_temp = 1;
			next;
		}
		if($check_temp) {
			if($data[$l] =~ /Gpu.*?(?:Current Temp)?/i) {
				my (undef, $tmp) = split(':', $data[$l]);
				if($tmp eq "\n") {
					$l++;
					$tmp = $data[$l];
				}
				my ($value, undef) = split(' ', $tmp);
				$value =~ s/[-]/./;
				$value =~ s/[^0-9.]//g;
				if(int($value) > 0) {
					$temp = int($value);
				}
			}
			$check_temp = 0;
		}
	}

	# NVIDIA driver v285.+ not supported (needs new output parsing).
	# This is to avoid a divide by zero message.
	if($total) {
		$mem = ($used * 100) / $total;
	} else {
		$mem = $used = $total = 0;
	}
	return join(" ", $mem, $cpu, $temp);
}

sub get_ati_data {
	my $myself = (caller(0))[3];
	my ($gpu) = @_;
	my $temp = 0;

	my @data = ();
	if(open(IN, "aticonfig --odgt --adapter=$gpu |")) {
		@data = <IN>;
		close(IN);
	} else {
		logger("$myself: ERROR: 'aticonfig' command is not installed.");
	}
	foreach(@data) {
		if(/Sensor \d: Temperature - (\d+\.\d+) C/) {
			$temp = $1;
		}
	}
	return $temp || 0;
}

# flushes out all Monitorix iptables/ipfw rules
sub flush_accounting_rules {
	my ($config, $debug) = @_;
	my $table = $config->{ip_default_table};

	if($config->{os} eq "Linux") {
		my $num = 0;
		my $num6 = 0;
		my $cmd = "iptables" . $config->{iptables_wait_lock};
		my $cmd6 = "ip6tables" . $config->{iptables_wait_lock};

		logger("Flushing out iptables rules.") if $debug;
		{
			my @names;

			# IPv4
			if(open(IN, "$cmd -t $table -nxvL INPUT --line-numbers |")) {
				my @rules;
				while(<IN>) {
					my ($rule, undef, undef, $name) = split(' ', $_);
					if(lc($config->{use_external_firewall} || "") eq "n") {
						if($name =~ /monitorix_IN/ || /monitorix_OUT/ || /monitorix_nginx_IN/) {
							push(@rules, $rule);
							push(@names, $name);
						}
					}
				}
				close(IN);
				@rules = reverse(@rules);
				foreach(@rules) {
					system("$cmd -t $table -D INPUT $_");
					$num++;
				}
			}
			if(open(IN, "$cmd -t $table -nxvL OUTPUT --line-numbers |")) {
				my @rules;
				while(<IN>) {
					my ($rule, undef, undef, $name) = split(' ', $_);
					if(lc($config->{use_external_firewall} || "") eq "n") {
						if($name =~ /monitorix_IN/ || /monitorix_OUT/ || /monitorix_nginx_IN/) {
							push(@rules, $rule);
						}
					}
				}
				close(IN);
				@rules = reverse(@rules);
				foreach(@rules) {
					system("$cmd -t $table -D OUTPUT $_");
					$num++;
				}
			}
			foreach(@names) {
				system("$cmd -t $table -X $_");
			}

			# IPv6
			if(lc($config->{ipv6_disabled} || "") ne "y") {
				undef(@names);
				if(open(IN, "$cmd6 -t $table -nxvL INPUT --line-numbers |")) {
					my @rules;
					while(<IN>) {
						my ($rule, undef, undef, $name) = split(' ', $_);
						if(lc($config->{use_external_firewall} || "") eq "n") {
							if($name =~ /monitorix_IN/ || /monitorix_OUT/ || /monitorix_nginx_IN/) {
								push(@rules, $rule);
								push(@names, $name);
							}
						}
					}
					close(IN);
					@rules = reverse(@rules);
					foreach(@rules) {
						system("$cmd6 -t $table -D INPUT $_");
						$num6++;
					}
				}
				if(open(IN, "$cmd6 -t $table -nxvL OUTPUT --line-numbers |")) {
					my @rules;
					while(<IN>) {
						my ($rule, undef, undef, $name) = split(' ', $_);
						if(lc($config->{use_external_firewall} || "") eq "n") {
							if($name =~ /monitorix_IN/ || /monitorix_OUT/ || /monitorix_nginx_IN/) {
								push(@rules, $rule);
							}
						}
					}
					close(IN);
					@rules = reverse(@rules);
					foreach(@rules) {
						system("$cmd6 -t $table -D OUTPUT $_");
						$num6++;
					}
				}
				foreach(@names) {
					system("$cmd6 -t $table -X $_");
				}
			}
		}
		if(open(IN, "$cmd -t $table -nxvL FORWARD --line-numbers |")) {
			my @rules;
			my @names;
			while(<IN>) {
				my ($rule, undef, undef, $name) = split(' ', $_);
				if($name =~ /monitorix_daily_/ || /monitorix_total_/) {
					push(@rules, $rule);
					push(@names, $name);
				}
			}
			close(IN);
			@rules = reverse(@rules);
			foreach(@rules) {
				system("$cmd -t $table -D FORWARD $_");
				$num++;
			}
			foreach(@names) {
				system("$cmd -t $table -F $_");
				system("$cmd -t $table -X $_");
			}
		}
		if(lc($config->{ipv6_disabled} || "") ne "y") {
			if(open(IN, "$cmd6 -t $table -nxvL FORWARD --line-numbers |")) {
				my @rules;
				my @names;
				while(<IN>) {
					my ($rule, undef, undef, $name) = split(' ', $_);
					if($name =~ /monitorix_daily_/ || /monitorix_total_/) {
						push(@rules, $rule);
						push(@names, $name);
					}
				}
				close(IN);
				@rules = reverse(@rules);
				foreach(@rules) {
					system("$cmd6 -t $table -D FORWARD $_");
					$num6++;
				}
				foreach(@names) {
					system("$cmd6 -t $table -F $_");
					system("$cmd6 -t $table -X $_");
				}
			}
		}
		logger("$num iptables rules have been flushed.") if $debug;
		if(lc($config->{ipv6_disabled} || "") ne "y") {
			logger("$num6 ip6tables rules have been flushed.") if $debug;
		}
	}
	if(grep {$_ eq $config->{os}} ("FreeBSD", "OpenBSD", "NetBSD")) {
		logger("Flushing out ipfw rules.") if $debug;
		system("ipfw delete $config->{port}->{rule} 2>/dev/null");
		system("ipfw delete $config->{nginx}->{rule} 2>/dev/null");
	}
}

sub lockfile_handler {
	my ($config) = @_;
	if (lc($config->{enable_rrd_lock} || "") eq "y") {
		my $lock_file = "/tmp/monitorix.lock";

		my $lockfile_was_available = (-e $lock_file);
		open(my $fh, ($lockfile_was_available ? "+<" : ">>"), $lock_file) or die "Can't open $lock_file: $!"; # If the file already exists we open it without the O_CREATE flag due to the limitations introduced by fs.protected_regular.
		if (!$lockfile_was_available) {
			chmod(0666, $lock_file);
		}
		return $fh;
	}
	return undef;
}

sub global_flock {
	my ($fh, $option) = @_;
	if (defined($fh)) {
		flock($fh, $option) or die "flock error: $!\n";
	}
}

1;
